designs/statefulset-disruption.md

```diff
@@ -0,0 +1,757 @@
+
+### Solutions:
+
+We should:
+- A1: Fix Kubelet race condition upstream for future Kubernetes versions
+- A2: Have Karpenter taint terminated nodes as `out-of-service` before removing its finalizer. 
+
+#### A1: Fix race at Kubelet level
+
+The Kubelet Shutdown Manager should not kill CSI Driver Pods before volumes are unmounted. This change must be made at `kubernetes/kubernetes` level.
+
+See: [Active PR #125070](https://github.com/kubernetes/kubernetes/pull/125070) for more information. 
+
+**Pros:** 
+
+- Other cluster autoscalers will not face this race condition.  
+- Ideal long-term solution
+- Reduces pod migration times
+- Reduces risk of data corruption
+
+**Cons:**
+
+- Unavailable until merged in a version of Kubernetes (Likely Kubernetes `v1.31`) (Possibly able to be backported)
+
+#### A2: Taint node as `out-of-service` after termination
+
+While this race should be fixed at the Kubelet level long-term, we still need a solution for earlier versions of Kubernetes.
+
+One solution is to mark terminated nodes via `out-of-service` taint. 
+
+In `v1.26`, Kubernetes enabled the [Non-graceful node shutdown handling feature](https://kubernetes.io/docs/concepts/cluster-administration/node-shutdown/#non-graceful-node-shutdown) by default. This introduces the `node.kubernetes.io/out-of-service` taint, which can be used to mark a node as terminated. 
+
+Once Karpenter confirms that an instance is terminated, adding this taint to the node object will allow the Attach/Detach Controller to treat the volume as not in use, preventing the 6+ minute delay. 
+
+By modifying Karpenter to apply this taint and wait ~5 seconds until volumes are marked not in use, the following sequence will occur:
+
+![taint_solution](https://github.com/AndrewSirenko/karpenter-provider-aws/assets/68304519/d1bc5704-7592-4aa1-a72e-83d33f824b8b)
+
+See [this commit](https://github.com/kubernetes-sigs/karpenter/pull/1294/commits/88134e00ad02863d5a7268bba0b639e21f3f5398) for a proof-of-concept implementation. 
+
+**Pros:**
+
+- Solves 6+ minute delays by default
+- No additional latency before starting instance termination. 
+- Minor latency in deleting Node's finalizer IF karpenter does not treat `shutting down` instance as terminated. (In my tests, a 5 second wait was sufficient for AttachDetach Controller to recognize the out-of-service taint and allow for volume detach)
+- If Kubernetes makes this 6 minute ForceDetach timer infinite by default (As currently planned in Kubernetes v1.32), the out-of-service taint will be the only ensure workload starts on other node 
+
+**Cons:**
+
+- Unavailable until Kubernetes `v1.26`. Customers running Karpenter on Kubernetes ≤ `v1.25` will require solution B1 to be implemented. 
+- Requires Karpenter-Provider-AWS to not treat `Shutting Down` as terminated (Though Karpenter had already planned on this via [PR #5979](https://github.com/aws/karpenter-provider-aws/pull/5979))
```

Comment:
I see #5979 was merged but only changed tests. Unsure what status is on this bullet point.